cobra/core/Metabolite.py

```diff
@@ -41,48 +43,57 @@ def __init__(self, id=None, formula=None,
+            self._formula = self.__dict__['formula']
+            return self._formula
+
+    @formula.setter
+    def formula(self, formula):
+        formula = str(formula) if formula is not None else ''
+        if "*" in formula:
+            warn("invalid character '*' found in formula '%s'" % formula)
+            formula = formula.replace("*", "")
+        if "(" in formula or ")" in formula:
+            warn("invalid formula (has parenthesis) in '%s'" % formula)
+        self._formula = formula
+
+    @property
     def elements(self):
-        tmp_formula = self.formula
-        if tmp_formula is None:
-            return {}
-        # necessary for some old pickles which use the deprecated
-        # Formula class
-        tmp_formula = str(self.formula)
-        # commonly occuring characters in incorrectly constructed formulas
-        if "*" in tmp_formula:
-            warn("invalid character '*' found in formula '%s'" % self.formula)
-            tmp_formula = tmp_formula.replace("*", "")
-        if "(" in tmp_formula or ")" in tmp_formula:
-            warn("invalid formula (has parenthesis) in '%s'" % self.formula)
-            return None
-        composition = {}
-        parsed = element_re.findall(tmp_formula)
-        for (element, count) in parsed:
-            if count == '':
-                count = 1
-            else:
-                try:
-                    count = float(count)
-                    int_count = int(count)
-                    if count == int_count:
-                        count = int_count
-                    else:
-                        warn("%s is not an integer (in formula %s)" %
-                             (count, self.formula))
-                except ValueError:
-                    warn("failed to parse %s (in formula %s)" %
-                         (count, self.formula))
-                    return None
-            if element in composition:
-                composition[element] += count
-            else:
-                composition[element] = count
-        return composition
+        """A dictionary breaking the chemical formula down by element."""
+
+        def int_or_float(s):
+            try:
+                return int(s)
+            except ValueError:
+                # Do we want float warnings?
+                # I don't, but they would go here.
+                return float(s)
+
+        return {element: int_or_float(stoich) if stoich else 1
+                for element, stoich in element_re.findall(self.formula)}
+
+    @elements.setter
+    def elements(self, elements_dict):
+
+        def formula_items():
+            for element, stoich in iteritems(elements_dict):
+                # Ignore stoichiometric entries close to machine precision.
+                if abs(stoich) < 1E-10:
+                    continue
+                yield ''.join((element, str(stoich) if stoich != 1 else ''))
+
+        self.formula = ''.join(formula_items())
```

Comment:
Wouldn't it be inefficient to set formula if it has already been parsed into elements? I would think you would want to do self._formula = ''.join(formula_items())